keps/sig-apps/1847-autoremove-statefulset-pvcs/README.md

```diff
@@ -0,0 +1,453 @@
+`patch` on PVC resource.
+
+### Test Plan
+
+1. Unit tests
+
+1. e2e tests
+    - DeleteOnScaleDown
+      1. Create 2 pod statefulset, scale to 1 pod, confirm PVC deleted
+      1. Create 2 pod statefulset, add data to PVs, scale to 1 pod, scale back to 2, confirm PV empty.
+      1. Create 2 pod statefulset, delete stateful set, confirm PVCs deleted.
+      1. Create 2 pod statefulset, add data to PVs, manually delete one pod, confirm pod comes back and PV still has data (PVC not deleted).
+      1. As above, but manually delete all pods in stateful set.
+      1. Create 2 pod statefulset, add data to PVs, manually delete one pod, immediately scale down to one pod, confirm PVC is deleted.
+      1. Create 2 pod statefulset, add data to PVs, manually delete one pod, immediately scale down to one pod, scale back to two pods, confirm PV is empty.
+      1. Create 2 pod statefulset, add data to PVs, perform rolling confirm PVC don't get deleted and PV contents remain intact and useful in the updated pods.
+    - DeleteOnStatefulSetDeletion
+      1. Create 2 pod statefulset, scale to 1 pod, confirm PVC still exists,
+      1. Create 2 pod statefulset, add data to PVs, scale to 1 pod, scale back to 2, confirm PV has data (PVC not deleted).
+      1. Create 2 pod statefulset, delete stateful set, confirm PVCs deleted
+      1. Create 2 pod statefulset, add data to PVs, manually delete one pod, confirm pod comes back and PV has data (PVC not deleted).
+      1. As above, but manually delete all pods in stateful set.
+      1. Create 2 pod statefulset, add data to PVs, manually delete one pod, immediately scale down to one pod, confirm PVC exists.
+      1. Create 2 pod statefulset, add data to PVs, manually delete one pod, immediately scale down to one pod, scale back to two pods, confirm PV has data.
+    - Retain: 
+      1. same tests as above, but PVCs not deleted in any case and confirm data intact on the PV.
+    - Pod restart tests:
+      1. Create statefulset, perform rolling update 
+1. Upgrade/Downgrade tests
+    1. Create statefulset in previous version and upgrade to the version 
+       supporting this feature. The PVCs should remain intact.
+    2. Downgrade to earlier version and check the PVCs with Retain
+       remain intact and the others with set policies before upgrade 
+       gets deleted based on if the references were already set.
+1. Feature disablement/enable test for alpha feature flag `statefulset-autodelete-pvcs`.
+
+
+### Graduation Criteria
+
+#### Alpha release
+- Complete adding the items in the 'Changes required' section.
+- Add unit, functional, upgrade and downgrade tests to automated k8s test.
+
+### Upgrade / Downgrade Strategy
+
+This features adds a new field to the StatefulSet. The default value for the new field
+maintains the existing behavior of StatefulSets.
+
+On a downgrade, the PersistentVolumeClaimReclaimPolicy field will be removed from any
```

Comment:
clarified

Reply:
In the deletion case, is it ok if we don't try to actively remove the ownerRefs if the feature gets disabled? The user turned off the feature, but they still end up with the feature behavior since we added the ownerRef while the feature was enabled.

We had another feature (storage protection) that had an issue like this, but the ramification there was that PVC deletion would hang after disablement/rollback, since there's nothing to operate on the finalizer anymore. We went and added logic to remove the finalizer when the feature is disabled.